[cse7761] Add Sonoff POW-CT support and frequency sensor

mick96 opened the pull request:
# What does this implement/fix?

- This PR implements the possibility to select the specific Sonoff device model to configure the cse7761 accordingly. This fixes issues on Sonoff POW Ring (POWCT).
Backward compatibility should be guaranteed by using default values on the new optional parameters.
- This PR implement optional frequency sensor available on cse7761.
- Improve code comments with cse7761 register details.

All the work is based on Tasmota implementation: https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xnrg_energy/xnrg_19_cse7761.ino

**All credit to @arendst and Tasmota community.**

## Types of changes

- [X] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**
- fixes https://community.home-assistant.io/t/sonoff-powct-pow-ring-power-not-accurate/789686

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#4469

**Pull request in [esphome-devices](https://github.com/esphome/esphome-devices) with documentation (if applicable):**

- esphome/esphome-devices#1007

## Test Environment

- [X] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
 sensor: 
  - platform: cse7761
    model: 'SONOFF POWCT'
    update_interval: 2s
    voltage:
      name: Voltage
    frequency:
      name: Frequency
    current_1:
      name: Current
    active_power_1:
      name: Power

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

mick96 committed:
Add Sonoff POW-CT support along with frequency sensor

Add configuration entry to set sonoff model:
  sonoff_model (Optional, int): Manually specify the Sonoff model, can be one of
  DUALR3, POWCT. Defaults to DUALR3.

This solve the following issue: https://community.home-assistant.io/t/sonoff-powct-pow-ring-power-not-accurate/789686

Add frequency sensor:
  frequency (Optional): Use the frequency value calculated by the meter. All options
  from Sensor.

mick96 committed:
Fix typeo in sensor.py

mick96 committed:
Clean up old comment

mick96 committed:
clang diff applied

mick96 committed:
Fix trailing spaces in sensor.py

mick96 commented:
When and if this PR will be merged the template for Sonoff POW Ring must be updated as well.
https://devices.esphome.io/devices/SONOFF-POWCT

mick96 committed:
Merge branch 'dev' into cse7761-sonoffpow-fix

mick96 committed:
Merge branch 'dev' into cse7761-sonoffpow-fix

tinusb commented:
Tested this PR and can confirm that it fixes the problem with the cse7761 platform and the new Sonoff POW-CT.  Without the fix the Sonoff POW-CT only provides accurate readings up to around 2000W.  Above 2000W cse7761 sensor output reduces to about 1/3 of the actual usage.  After I added the PR to my yaml script the cse7761 sensor output value was correct and I verified the readings with an amp clamp meter.  The sensor also outputs the mains frequency which is a handy feature.

lightingman117 commented:
Would love to see this merged when possible, thanks Devs!

333robbie333 commented:
I'd also like to see this merged. Thanks Devs!

mick96 committed:
Merge branch 'dev' into cse7761-sonoffpow-fix

h1771t commented:
This PR fixes all my POWCT issues. Applied the patches, added to my config 
`sonoff_model: POWCT`
Now the sensor readings to HA looks good below and above 2kw load, and over 24 hours the HA daily usage matches my utility usage reading very closely.

I removed 2 of the below config as well (for current_1 and for active_power_1), initially take from https://devices.esphome.io/devices/SONOFF-POWCT

> filters:
        # Measurement divided by the PI number
        - lambda: return x / PI;

So I think a patch needs to also go with this PR to remove the filters from the config documentation at https://devices.esphome.io/devices/SONOFF-POWCT

h1771t commented:
Would also love for this to get merged, cc @jesserockz

mick96 commented:
> This PR fixes all my POWCT issues. Applied the patches, added to my config `sonoff_model: POWCT` Now the sensor readings to HA looks good below and above 2kw load, and over 24 hours the HA daily usage matches my utility usage reading very closely.
> 
> I removed 2 of the below config as well (for current_1 and for active_power_1), initially take from https://devices.esphome.io/devices/SONOFF-POWCT
> 
> > filters:
> > # Measurement divided by the PI number
> > - lambda: return x / PI;
> 
> So I think a patch needs to also go with this PR to remove the filters from the config documentation at https://devices.esphome.io/devices/SONOFF-POWCT

Yeah, the division by PI is no longer needed and when and if this PR  will be merged the config documentation for POW-CT must be updated as well.
BTW the filter was a workaround for an incorrect CSE7761 register configuration, which this PR now corrects directly.

mick96 committed:
Merge branch 'dev' into cse7761-sonoffpow-fix

mick96 committed:
Merge branch 'dev' into cse7761-sonoffpow-fix

h1771t commented:
Is this waiting on @berfenger @jesserockz ? The PR fixes work great for me since putting it in, would be nice to have it committed. Thanks.

h1771t commented:
I would like to upgrade esphome on my devices, but this one is holding me from doing so, it will break one of my main devices I depend on. Any word on merging @jesserockz ? Thanks

lightingman117 commented:
> I would like to upgrade esphome on my devices, but this one is holding me from doing so, it will break one of my main devices I depend on. Any word on merging @jesserockz ? Thanks

Just do this, don't need to wait.

```
external_components:
  - source:
      type: git
      url: https://github.com/mick96/esphome.git
      ref: cse7761-sonoffpow-fix
    components: [ cse7761 ]
```

clydebarrow committed:
Merge branch 'dev' into cse7761-sonoffpow-fix

TimoPtr commented:
@jesserockz I tested the implementation and it works on my end on my device.

mick96 committed:
Merge branch 'dev' into cse7761-sonoffpow-fix

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

jesserockz commented on esphome/components/cse7761/cse7761.cpp:
```diff
@@ -216,6 +392,10 @@ void CSE7761Component::get_data_() {
     this->voltage_sensor_->publish_state(voltage);
   }
 
+  float freq = (this->data_.frequency) ? ((float) CSE7761_FREF / 8 / this->data_.frequency) : 0;  // Hz
+  if (this->frequency_sensor_ != nullptr) {
```
No need to do maths if not going to be used
```suggestion
  if (this->frequency_sensor_ != nullptr) {
    float freq = (this->data_.frequency) ? ((float) CSE7761_FREF / 8 / this->data_.frequency) : 0;  // Hz
```

jesserockz committed:
Update esphome/components/cse7761/cse7761.cpp

jesserockz committed:
Rename to model, add tests

jesserockz committed:
Fix

mick96 committed:
Merge branch 'esphome:dev' into cse7761-sonoffpow-fix

mick96 committed:
Use DEVICE_CLASS_FREQUENCY for frequency sensor

Set device_class=DEVICE_CLASS_FREQUENCY and remove icon override.

Correct device_class lets the HA UI know the default unit and icon and gives the sensor proper semantic meaning.

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

TimoPtr commented on esphome/components/cse7761/cse7761.cpp:
```diff
@@ -12,30 +12,52 @@ static const char *const TAG = "cse7761";
+// static const uint8_t CSE7761_REG_POWERPBC       = 0x74;     // (2) Channel B active power conversion coefficient
+// static const uint8_t CSE7761_REG_POWERSC        = 0x75;     // (2) Apparent power conversion coefficient
+// static const uint8_t CSE7761_REG_ENERGYAC       = 0x76;     // (2) Channel A energy conversion coefficient
+// static const uint8_t CSE7761_REG_ENERGYBC       = 0x77;     // (2) Channel B energy conversion coefficient
```
Why all of these const are commented?

mick96 replied:
These register definitions (0x71–0x77) are kept as comments for documentation/reference purposes. They're not used directly in code because chip_init_() reads all 8 coefficients in a loop using CSE7761_REG_RMSIAC + i (register 0x70 through 0x77). Only CSE7761_REG_RMSIAC (0x70) is needed as the base address.

swoboda1337 replied:
Should remove them or uncomment them.

mick96 replied:
If we uncomment them we'll get unused symbol warnings in CI, since they're not referenced in the current implementation.

swoboda1337 replied:
Make them constexpr instead dont think CI will complain.

TimoPtr commented on esphome/components/cse7761/cse7761.cpp:
```diff
@@ -149,13 +171,17 @@ uint32_t CSE7761Component::read_(uint8_t reg, uint8_t size) {
 uint32_t CSE7761Component::coefficient_by_unit_(uint32_t unit) {
+  uint32_t coeff = 1;
+  if (this->data_.model == CSE7761_MODEL_SONOFF_POWCT) {
+    coeff = 5;
```
Why 5? It would be nice to have some documentation for it.

mick96 replied:
The Dual R3 uses a shunt resistor with PGA=16×, while the POW CT uses a current transformer with PGA=1× (higher gain would clip the ADC). The chip's calibration coefficients are factory-burned by the supplier and identical across both products. The factor of 5 compensates for the net analog front-end gain difference. 
This value is taken from the Tasmota implementation and confirmed working.
I'll add a comment.

TimoPtr commented on esphome/components/cse7761/cse7761.cpp:
```diff
@@ -179,10 +205,157 @@ bool CSE7761Component::chip_init_() {
+      1111   Current channel B overcurrent signal indication output
+    */
+
+    // this->write_(CSE7761_REG_PULSE1SEL | 0x80, 0x3290);  // Enable zero crosing signal output on function pin
```
dead code?

mick96 replied:
This is intentionally commented out. It would configure the PULSE1 pin to output the voltage zero-crossing signal (P1Sel=0b1001 = 0x9, visible in the register table above it). It's left as a reference for anyone who wants to enable zero-crossing detection on the function pin.

swoboda1337 commented on esphome/components/cse7761/cse7761.cpp:
```diff
@@ -12,30 +12,52 @@ static const char *const TAG = "cse7761";
+//                                                             overload is channel B
 static const uint8_t CSE7761_CMD_CLOSE_WRITE = 0xDC;   // Close write operation
 static const uint8_t CSE7761_CMD_ENABLE_WRITE = 0xE5;  // Enable write operation
 
```
Same here, remove commented out register/command definitions

mick96 replied:
I'd prefer to keep them as comments for reference. These are register and command definitions from the CSE7761 datasheet — they document chip capabilities that aren't actively used in the current implementation but are useful context for anyone working on this driver (e.g., adding apparent power, energy accumulation, or channel B selection).

swoboda1337 replied:
Then uncomment them. It wont get merged like this.

swoboda1337 commented on esphome/components/cse7761/cse7761.cpp:
```diff
@@ -179,10 +205,157 @@ bool CSE7761Component::chip_init_() {
+      R3 / Pow CT) 0      NC                 Default is 1
+    */
+
+    this->write_(CSE7761_REG_EMUCON2 | 0x80, 0x0FE5);  // Sonoff Dual R3 / Pow CT + frequency measure enable
```
This changed from `0x0FC1` to `0x0FE5` which enables zero-crossing (bit 2) and waveform output (bit 5) for all models, not just POWCT. Could this have side effects for existing Dual R3 users?

mick96 replied:
These are required for the frequency sensor to work — the UFREQ register (0x23) only produces valid data when zero-crossing detection is enabled, and ZxEN requires WaveEN as a prerequisite per the datasheet.

Regarding side effects for Dual R3 users: enabling these bits only activates internal measurement modules inside the chip. They don't change how voltage, current, or power are calculated — those paths are independent. If a Dual R3 user doesn't configure a frequency sensor, the value is read but never published, so there's no functional change for existing configurations.

swoboda1337 commented on esphome/components/cse7761/cse7761.cpp:
```diff
@@ -179,10 +205,157 @@ bool CSE7761Component::chip_init_() {
+      1101   Voltage channel undervoltage indication signal output
+      1110   Current channel A overcurrent signal indication output
+      1111   Current channel B overcurrent signal indication output
+    */
```
The PULSE1SEL write was removed — old code wrote `0x3290`, now its commented out. This changes default behavior for existing Dual R3 users. Is this intentional?

mick96 replied:
Intentional. The 0x3290 value only changed what signal appears on the physical Pulse2 pin (zero-crossing instead of default PFB pulse). ESPHome doesn't monitor that pin, so it has no functional impact. Default 0x3210 is fine. Kept as a comment for reference.

Could you clarify the project's guidelines on commented-out code?

swoboda1337 commented on esphome/components/cse7761/cse7761.cpp:
```diff
@@ -179,10 +205,157 @@ bool CSE7761Component::chip_init_() {
+      PFA pulse output and active energy register accumulation; (Sonoff Dual R3 / Pow CT) =0 (default), turn off PFA
+      pulse output and active energy register accumulation.
+    */
+
```
These register bit-field descriptions are copy-pasted datasheet content. A brief comment explaining the value chosen (e.g. `// PGA=1 for all channels, channel B enabled`) with a datasheet reference would be much cleaner than 100+ lines of register docs

mick96 replied:
Ok, but consider that the only datasheet I could find is in Chinese.

swoboda1337 commented on esphome/components/cse7761/cse7761.cpp:
```diff
@@ -216,6 +392,10 @@ void CSE7761Component::get_data_() {
 
+  if (this->frequency_sensor_ != nullptr) {
+    float freq = (this->data_.frequency) ? ((float) CSE7761_FREF / 8 / this->data_.frequency) : 0;  // Hz
+    this->frequency_sensor_->publish_state(freq);
```
nit: missing blank line before the for loop

mick96 replied:
I'll add it

mick96 commented on esphome/components/cse7761/sensor.py:
```diff
@@ -60,6 +70,15 @@
+                device_class=DEVICE_CLASS_FREQUENCY,
+                state_class=STATE_CLASS_MEASUREMENT,
+            ),
+            cv.Optional(CONF_MODEL, default="SONOFF DUALR3"): cv.enum(
+                MODELS, upper=True
+            ),
```
PR description updated

mick96 committed:
refactor(cse7761): Use constexpr and clean up register definitions

- Replace `static const` with `static constexpr` for compile-time constant evaluation
- Uncomment previously disabled register and command definitions to make them available
- Fix data comparison logic from `<` to `<=` for proper 0xFF boundary handling
- Simplify multi-line comments for channel selection commands

mick96 committed:
Merge branch 'dev' into cse7761-sonoffpow-fix

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

mick96 committed:
fix(cse7761): Fix register definition alignment and code formatting

- Remove erroneous closing brace in write_() function
- Fix malformed loop in read_once_() function with proper CRC check logic

mick96 committed:
Merge branch 'cse7761-sonoffpow-fix' of https://github.com/mick96/esphome into cse7761-sonoffpow-fix

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

mick96 committed:
fix(cse7761): fix messy merge

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

mick96 committed:
feat(cse7761): Add get_setup_priority method declaration

mick96 committed:
Merge branch 'cse7761-sonoffpow-fix' of https://github.com/mick96/esphome into cse7761-sonoffpow-fix

mick96 committed:
Merge branch 'dev' into cse7761-sonoffpow-fix

mick96 committed:
Merge branch 'dev' into cse7761-sonoffpow-fix

mick96 committed:
Merge branch 'dev' into cse7761-sonoffpow-fix
